### ISSUE TITLE
inkompatible with new Unifi key

### ISSUE THREAD

**Issue by mrmorgul** [oznu/homebridge-unifi-occupancy-sensor#27]
**Describe The Bug:**
After upgrade unifi cloudkey+ to firmware 2.0.24 homebridge is unable to login to Unifi.
Network version 6.0.43 latest version, this was not updated. and worked with previous cloudkey firmware 1.11.3
Protect version 1.16.9 works with "Homebridge UniFi Protect" with same user and password.

No changes is made to homebridge settings.

User scheme in cloudkey is recreated with same username and password, as Unifi have changes in this area.

as cloudkey on new firmware is pretty slow, I reconfigured the entire unifi setup, with same result.

**To Reproduce:**

Upgrade Unifi Cloudkey to latest stable (?) firmware

https://192.168.1.2:8443/manage/site/default/dashboard - return "Connection refused"
https://192.168.1.2/manage/site/default/dashboard - redirects to cloudkey home screen
https://192.168.1.2/network/site/default/dashboard - return "OK"

**Expected behavior:**
unable to login

**Logs:**

```
[26.12.2020 15.10.23] [TV_On] UniFi Events: Reconnecting (224)
[26.12.2020 15.10.23] [TV_On] UniFi Events: Login Failed Error: connect ECONNREFUSED 192.168.1.2:8443
[26.12.2020 15.10.23] [TV_On] UniFi Events: Error (225) - connect ECONNREFUSED 192.168.1.2:8443
[26.12.2020 15.10.23] [TV_On] UniFi Events: Disconnected (225) - Retry in 5000ms
```
<After removing ":8443" from config:>
<"controller": "https://192.168.1.2">
[26.12.2020 15.33.51] [TV_On] UniFi Events: Reconnecting (6)
[26.12.2020 15.33.51] [TV_On] UniFi Events: Connected (7) - wss://192.168.1.2/wss/s/default/events
[26.12.2020 15.33.51] [TV_On] UniFi Events: Disconnected (7) - Retry in 5000ms

**Homebridge Config:**

```json
        {
            "accessory": "UniFi Occupancy Sensor",
            "name": "TV_On",
            "mode": "any",
            "interval": 120,
            "unifi": {
                "controller": "https://192.168.1.2:8443",
                "username": "xxxxx",
                "password": "xxxxx",
                "site": "default"
            },
            "monitor": [
                {
                    "device": "c0:48:e6:8e:48:05"
                }
            ]
        },
```

**Screenshots:**

**Environment:**

* **Node.js Version**: 
* **NPM Version**: 
* **Homebridge Version**:  
* **Homebridge Plugin Version**: homebridge-unifi-occupancy-sensor v1.0.0
* **Homebridge Config UI X Plugin Version**: homebridge-config-ui-x v4.36.0
* **Operating System**:  Raspian PI

**Comment by gussyboy** [comment]
What am I missing, how does this package get updated? @tickez ??

**Comment by twynne** [comment]
I wonder if the patch mentioned here [https://github.com/oznu/homebridge-unifi-occupancy-sensor/issues/20](url) would fix it?  It's for the UDM on Unifi OS, and I assume the upgrade to Unifi OS on CGK2+ is the source of the issues.

**Comment by tickez** [comment]
I raised PR #28 to enable support for both UnifiOS and "legacy" devices

**Comment by twynne** [comment]
How does this now get published?

**Comment by tickez** [comment]
It depends on the original author of the plugin to accept the change and release a new version if everything is in order.
If you know what you're doing and want to give it a try, you can install the plugin from my repo: 
npm install "tickez/homebridge-unifi-occupancy-sensor#unifiOS"

**Comment by angrest** [comment]
when I install from your repo @tickez I get an error message in the log: [Occupancy Sensor] ERROR: Failed to check occupancy: Cannot read property 'length' of undefined

**Comment by mrmorgul** [comment]
> Thanks @tickez for the update - yesterday I upgraded to 6.0.43 and suddenly, the plugin stopped working.
> It seems that the GUI for the settings does not yet reflect the new parameter "unifios" and thus, I got an error message on first try.
> Now, I get an error message
> `ERROR: Failed to check occupancy: 401 - "Unauthorized"`
> It seems the plugin works anyway, so this is obviously not fatal.

Check cloudkey users, they have changed the method, after i recreated the user i use for homebridge, i solved the unautbrized error, and get ECONNREFUSED instead.

**Comment by tickez** [comment]
I unfortunately have no way to test this as I only run a standalone Controller 6.0.43, but it is important that you set `unifios: true` and use port 443 instead of 8443 if you're on a CloudKey or UDM.

**Comment by angrest** [comment]
> > Thanks @tickez for the update - yesterday I upgraded to 6.0.43 and suddenly, the plugin stopped working.
> > It seems that the GUI for the settings does not yet reflect the new parameter "unifios" and thus, I got an error message on first try.
> > Now, I get an error message
> > `ERROR: Failed to check occupancy: 401 - "Unauthorized"`
> > It seems the plugin works anyway, so this is obviously not fatal.
> 
> Check cloudkey users, they have changed the method, after i recreated the user i use for homebridge, i solved the unautbrized error, and get ECONNREFUSED instead.

And this error is preferable? I'm confused...

**Comment by angrest** [comment]
> I unfortunately have no way to test this as I only run a standalone Controller 6.0.43, but it is important that you set `unifios: true` and use port 443 instead of 8443 if you're on a CloudKey or UDM.

Did both.

Happy to help if I can do some tests for you (or send more information)

**Comment by tickez** [comment]
@angrest, can you confirm: Your Controller is hosted on a CloudKey or UDM, you've installed the plugin from my forked git repo and you're getting 401 - Unauthorized?
Is your account a UI SSO account and is it is, is 2FA disabled?
Does any of the "Corner Cases" mentioned on this page apply to you? https://community.ui.com/releases/UniFi-Cloud-Key-Firmware-2-0-24/b6684f1e-8542-4660-bc0b-74e0634448e8

**Comment by angrest** [comment]
@tickez 
* yes, my controller is hosted on a Cloud Key (Gen2)
* Yes, it's installed from your fork 
* Yes, I do get "unauthorized" on the first connection attempt. Thereafter, it seems to work.
* The account used is a local account without 2FA and without email-Address, so I suppose the corner cases mentioned do not apply.

**Comment by FugiOnline** [comment]
@oznu could you push out a new version? I'm also having issues with the new cloud key firmware version and it would be great to keep using this plugin.

**Comment by FugiOnline** [comment]
> I wonder if the patch mentioned here [https://github.com/oznu/homebridge-unifi-occupancy-sensor/issues/20](url) would fix it? It's for the UDM on Unifi OS, and I assume the upgrade to Unifi OS on CGK2+ is the source of the issues.

I can confirm that the patch mentioned [here](https://github.com/oznu/homebridge-unifi-occupancy-sensor/issues/20) makes the plugin work with the updated CloudKey Gen2 (UnifiOS). Do note I had to manually update tough-cookie as well.

**Comment by oznu** [comment]
Should be resolved in v2.0.0.

https://github.com/oznu/homebridge-unifi-occupancy-sensor/pull/28
